src/reducers/importText.ts

```diff
@@ -199,30 +199,49 @@ const importText = (state: State, { path, text, lastUpdated, preventSetCursor, r
 
     const uuid = createId()
 
+    const isDestContextEmpty = getAllChildren(state, context).length === 0
+
+    /** Check if destination's parent context has more than one children. */
+    const isDestParentContextEmpty = () => getAllChildren(state, rootedParentOf(state, context)).length <= 1
+
+    const destEmpty = destThought.value === '' && isDestContextEmpty
+
+    const noNeedOfCollapse = (!destEmpty && isDestContextEmpty) || (destEmpty && isDestParentContextEmpty())
+
     // Note: Create a dummy thought and then import new thoughts into its context and then collpase it . Since collapse uses existingThoughtMove it merges imported thoughts.
-    const updatedState = newThought(state, {
+    const updatedState = noNeedOfCollapse ? state : newThought(state, {
       at: simplePath,
       insertNewSubthought: true,
       value: uuid
     })
 
-    const destEmpty = destThought.value === '' && getAllChildren(state, context).length === 0
-
-    // New dummy thoughts for collapsing
-    const newDummyThought = getAllChildren(updatedState, context).find(child => child.value === uuid)
+    /**
+     * Returns destination path.
+     */
+    const getDestinationPath = () => {
+      if (noNeedOfCollapse) return simplePath
+      const newDummyThought = getAllChildren(updatedState, context).find(child => child.value === uuid)
+      return (newDummyThought ? [...simplePath, newDummyThought] : simplePath) as SimplePath
+    }
 
-    const newDestinationPath = unroot(newDummyThought ? [...simplePath, newDummyThought] : simplePath) as SimplePath
+    const newDestinationPath = getDestinationPath()
 
     const imported = importJSON(updatedState, newDestinationPath, json, { lastUpdated, skipRoot })
 
     /** Set cursor to the last imported path. */
     const setLastImportedCursor = (state: State) => {
 
       const lastImportedContext = pathToContext(imported.lastImported)
-      const cursorContextHead = lastImportedContext.slice(0, newDestinationPath.length - (destEmpty ? 2 : 1))
-      const cursorContextTail = lastImportedContext.slice(newDestinationPath.length)
-      const newCursorContext = cursorContextHead.concat(cursorContextTail)
-      const newCursor = rankThoughtsFirstMatch(state, newCursorContext)
+
+      /** Get last iumported cursor after using collapse. */
+      const getLastImportedAfterCollpase = () => {
+        const cursorContextHead = lastImportedContext.slice(0, newDestinationPath.length - (destEmpty ? 2 : 1))
+        const cursorContextTail = lastImportedContext.slice(newDestinationPath.length)
+        const newCursorContext = cursorContextHead.concat(cursorContextTail)
```

Comment:
I recommend getting in the habit of `[...cursorContextHead, ...cursorContextTail]` as its cleaner and faster.